connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java

```diff
@@ -28,4 +30,46 @@
     protected SourceConnectorContext context() {
         return (SourceConnectorContext) context;
     }
+
+    /**
+     * Signals whether the connector supports exactly-once delivery guarantees with a proposed configuration.
+     * Developers can assume that worker-level exactly-once support is enabled when this method is invoked.
+     *
+     * <p>For backwards compatibility, the default implementation will return {@code null}, but connector developers are
+     * strongly encouraged to override this method to return a non-null value such as
+     * {@link ExactlyOnceSupport#SUPPORTED SUPPORTED} or {@link ExactlyOnceSupport#UNSUPPORTED UNSUPPORTED}.
+     *
+     * <p>Similar to {@link #validate(Map) validate}, this method may be called by the runtime before the
+     * {@link #start(Map) start} method is invoked when the connector will be run with exactly-once support.
+     *
+     * @param connectorConfig the configuration that will be used for the connector.
+     * @return {@link ExactlyOnceSupport#SUPPORTED} if the connector can provide exactly-once support with the given
+     * configuration, and {@link ExactlyOnceSupport#UNSUPPORTED} if it cannot. If this method is overridden by a
+     * connector, should not be {@code null}, but if {@code null}, it will be assumed that the connector cannot provide
+     * exactly-once guarantees.
```

Comment:
I don't think we need this sentence: `If this method is overridden by a connector...`, because we still allow `null` return value, right? I think we just need to put the 3 kinds of return value clear here: `supported`, `unsupported`, `null`. WDYT?

Reply:
We do technically allow `null`, but we discourage it. This allows us to differentiate between a connector that has implemented this method and explicitly does not provide exactly-once support (signaled by returning `ExactlyOnceSupport.UNSUPPORTED`), and a connector that hasn't implemented this method at all (signaled by returning `null`). This comes into play in the downstream [preflight validation PR](https://github.com/apache/kafka/pull/11776):

https://github.com/C0urante/kafka/blob/b4976e82aec1c864bd25d660b4042a71e03b7c47/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L892-L899

Reply:
I see. Thanks.